### ISSUE TITLE
Adaptive Concurrency Control L7 Filter

### ISSUE THREAD

**Issue by tonya11en** [envoyproxy/envoy#7789]
Adaptive Concurrency Control HTTP Filter
---------------------------------------------------------

*TL; DR:
L7 filter to dynamically update request concurrency values based on sampled request latencies.* 

### Motivation
At Lyft, we’re using Envoy’s circuit breaking system to defend our services from load caused by too many concurrent requests. The circuit breakers currently require service owners to manually set limits for each service/upstream cluster. While the feature is beneficial, the service owners are expected to understand the ideal concurrency limits of their services, configure them, and update the limits as their service evolves. This is not a problem if the concurrency values are always tuned properly, but the inevitable stale values have caused several incidents in production.

This work aims to remove the need for manual tuning of these circuit breaking parameters and keep latencies low by using techniques from [Netflix’s concurrency limits Java library](https://github.com/Netflix/concurrency-limits).

### Planned Patches
The filter will be committed in stages to make the review process easier. Here's an ordered list of what is currently planned:

- [x] L7 filter with calls to a noop concurrency control framework.
- [x] Implement concurrency control framework and unit tests.
- [x] Make parameters runtime configurable.
- [x] Integration tests.
- [ ] Documentation.

Nice-to-haves:
- [ ] Benchmark.

### Preliminary Data
To test the various concurrency control algorithms, I used [bufferbloater](https://github.com/tonya11en/bufferbloater) (whose functionality will eventually be added to Nighthawk) and a proof-of-concept implementation in my [envoy-filter-example fork](https://github.com/tonya11en/envoy-filter-example/tree/acc_poc).

Identical tests were performed on a system with sane CB settings configured and adaptive concurrency configured. The client RPS to the server remained constant, but the server's response times varied every 30 seconds.

#### Circuit breaker configured
![realistic_cb_well_configured](https://user-images.githubusercontent.com/1109246/62255847-deccea80-b3b3-11e9-9700-b14a754c7743.png)

#### Adaptive concurrency control configured
![realistic_server_degrade](https://user-images.githubusercontent.com/1109246/62255856-e8eee900-b3b3-11e9-99db-9ffd26d5702f.png)

So far, the results seem promising. The proof-of-concept filter was able to replicate the circuit breaker behavior, but did so with latencies that were <= the CB test latencies. All while not requiring any manual concurrency configurations.

**Comment by mattklein123** [comment]
This plans LGTM to me @tonya11en. Very excited to see this land. In the proposed patch set:

- L7 filter with calls to a noop concurrency control framework.
- Implement concurrency control framework and unit tests.
- Make parameters runtime configurable.
- Integration tests and benchmark.
- Documentation.

In which PR is the real algorithm going to be wired into the filter? The integration test PR? If so that makes sense to me. Also, you might want to split stats out into its own PR and do that after the first PR with the NOP filter, but up to you.

**Comment by tonya11en** [comment]
@mattklein123 I plan to implement the actual algorithm (~gradient2 from netflix's lib) in the 2nd PR (implementation of the framework).

**Comment by mattklein123** [comment]
OK. Since the algorithm should be behind an interface, I would prefer to see a dedicated algorithm PR with tests, and then we can do a further PR to actually integrate the real algorithm into the filter along with integration tests.

**Comment by tonya11en** [comment]
Keepalive.

**Comment by mmckeen** [comment]
Would there be any chance in this proposal to allow concurrency control at a finer level?  Maybe some form of client_id passed from callers that Envoy could maintain separate concurrency limits for?  There are sets of problems that an approach similar to client side throttling in https://landing.google.com/sre/sre-book/chapters/handling-overload/ might be able to help solve.

**Comment by mattklein123** [comment]
@mmckeen I've been discussing this offline with @tonya11en. I think we do want to allow concurrency to be set at a finer level either via upstream cluster, arbitrary header (e.g. client id), etc. We can discsus this once we land the initial support.

**Comment by tonya11en** [comment]
@mmckeen in addition to what Matt said, what you mention is an approach for implementing QoS using this concurrency control filter. This is under consideration as a next step and I had some thoughts on this - I'll CC you on the brainstorming doc when the time comes if you'd like.

**Comment by tonya11en** [comment]
@PiotrSikora there's currently a necessary default maximum concurrency in this filter's config proto. In steady-state, the concurrency limit continues to increase until there is a measurable increase in latency.

This might cause more problems for you with regards to https://github.com/istio/istio/issues/17233.

**Comment by PiotrSikora** [comment]
@tonya11en thanks!

**Comment by tonya11en** [comment]
Just to provide an update on what we have so far, I wanted to share some experimental data. Once we merge #8524, we'll have an MVP for the adaptive concurrency filter. It's behaving as expected with the bugfix.

Below you'll find some output from a client/server load generator that I hacked together. It's a 160s test with RPS slowly ramping up from to values that tip the fake server over. The fake server injects latencies of p99=150ms and p90=50ms with 8 worker threads. The fake client will time out a request at 1s.

### Filter passthrough, no concurrency limiting
Below is the output of a test run without circuit breaking or adaptive concurrency. It's just a simple pass-through to compare against.
![prod_bypass_filter](https://user-images.githubusercontent.com/1109246/66358810-4dd13d00-e929-11e9-8db4-039bf4cc49d3.png)

### Adaptive concurrency
The adaptive concurrency filter appears to keep the latencies relatively close to the ideal. The filter deduces the ideal concurrency (we know this to be ~8 based on the number of workers) purely based on latency samples.
![prod_with_filter](https://user-images.githubusercontent.com/1109246/66359023-38a8de00-e92a-11e9-85e7-2695e7b051ee.png)

Things look promising. I'll be writing up some thorough documentation once the remaining 2 patches are merged, then I'll see how it performs in our staging environment within the next few weeks. If anyone would like to assist in hardening the filter, I'd be happy to work with you.
